anndata/_core/sparse_dataset.py

```diff
@@ -474,6 +474,16 @@ def append(self, sparse_matrix: ss.spmatrix):
                 f"Matrices must have same format. Currently are "
                 f"{self.format!r} and {get_format(sparse_matrix)!r}"
             )
+        indptr_offset = self.group["indptr"][-1]
+        if indptr_offset.dtype == np.int32:
+            # This is difficult to test without requiring ~ 16GB of memory
+            new_nnz = int(indptr_offset) + int(sparse_matrix.indptr[-1])
+            if new_nnz >= np.iinfo(np.int32).max:
+                raise OverflowError(
+                    "This array was written with a 32 bit intptr, but is now large "
+                    "enough to require 64 bit values. Please recreate the array with "
+                    "a 64 bit indptr."
```

Comment:
Make clear how? In the release note?

Reply:
In this error -  you say "Please recreate the array with..." but it might be worth noting how one can do this (simply by re-writing the data with this later/st version of the package.

Reply:
That will only happen if they are using dask.

Reply:
Ok but what does that to do with letting people know that they can just re-run their writing code with this new version to fix the problem?

Reply:
I don't see a case where this would be possible.

Two things are changing in this PR.

* Writing a dask array with sparse chunks now won't have overflow-based data corruption errors (unless you start to need >64bit)
* If you attempt to append to an existing on disk sparse array, and that would cause an overflow-based corruption, you instead get an error message.

There is no case where you are appending to an existing dataset with dask arrays AFAIK.

Reply:
@ivirshup @ilan-gold what is the correct way to cast a sparse dataset to having indptrs that are int64? I am trying to combine ~20 million cells with ~20000 genes spread evenly across 100 h5ad files into a single h5ad file using `anndata.experimental.concat_on_disk` and seeing this `OverflowError`

- [ ]

Reply:
As a workaround, I've written a function that reads an anndata object, casts the indptr dtype to int64, and saves a new h5ad file, but it would be nice if AnnData just recognized that an overflow is going to occur in `concat_on_disk` and cast the type of the indptr on the fly

```
def convert_h5ad_sparse_matrix_indices_to_int64(h5ad_file, new_file):
    adata = ad.read_h5ad(h5ad_file)
    adata.X.indptr = np.array(adata.X.indptr, dtype=np.int64)
    adata.write_h5ad(new_file)
```

Reply:
@ivirshup was there a reason not to re-write in this function the `indptr`?

One way to work around this that would be more efficient would be to only open/write the `indptr` i.e., use `h5py` directly to do the above, changing only the dtype.  This would probably be much faster.  Something like:

```python
import h5py
import numpy as np
f = h5py.File('my_anndata_object.h5ad', 'r+')
indptr_int_64 =  np.array(f['X/indptr'], dtype='int64')
del f['X/indptr']
f['X/indptr'] = indptr_int_64
```

but maybe there's a reason not to do this that I'm missing?

Reply:
I mean, we don't want to rewrite only the `indptr` since that gives us problems downstream.

I think `concat_on_disk` could be modified to write `int64` `indptr` and `indices` by default.

@agdenadel, could you open this as an issue? Makes it easier for us to track in our workflow that way.

Reply:
I created #1470